File: sslib/validate.py
"""JSON validation for team data."""

from .constants import (
    CHARSET, CATEGORIES,
    COLOUR_VALUES, STYLE_VALUES, HEAD_VALUES, ROLE_VALUES,
    POSITION_VALUES, POSITION_NAMES, TACTIC_VALUES,
)
from .decode import find_pointer_table, chain_walk_region


def validate_string(text, context):
    """Check all characters are in CHARSET. Returns list of bad chars or empty list."""
    bad = []
    for c in text.upper():
        if c not in CHARSET:
            bad.append(c)
    return bad


def _check_enum(val, allowed, max_int, context):
    """Validate val is a known string or int in 0..max_int. Returns error or None."""
    if isinstance(val, str):
        if val not in allowed:
            return f"{context} must be one of {sorted(allowed)}, got '{val}'"
    elif not (0 <= val <= max_int):
        return f"{context} must be 0-{max_int}, got {val}"
    return None


def validate_teams(rom_bytes, teams_json):
    """Validate teams JSON against the ROM structure.

    Args:
        rom_bytes: ROM file contents (bytes or bytearray)
        teams_json: parsed JSON dict with 'national', 'club', 'custom' keys

    Returns:
        (errors, warnings) — both are lists of strings.
    """
    errors = []
    warnings = []

    if not isinstance(teams_json, dict) or not all(k in teams_json for k in CATEGORIES):
        errors.append("JSON must be a dict with 'national', 'club', 'custom' keys")
        return errors, warnings

    ptrs = find_pointer_table(rom_bytes)
    region_info = [
        ('national', ptrs['nat_start'], ptrs['nat_end']),
        ('club', ptrs['club_start'], ptrs['club_end']),
        ('custom', ptrs['cust_start'], ptrs['cust_end']),
    ]

    all_block_offsets = {}
    for cat, start, end in region_info:
        all_block_offsets[cat] = chain_walk_region(rom_bytes, start, end)

    for cat, start, end in region_info:
        rom_count = len(all_block_offsets[cat])
        json_teams = teams_json[cat]
        if len(json_teams) != rom_count:
            errors.append(f"{cat}: expected {rom_count} teams in JSON, got {len(json_teams)}")
            continue
        for i, team in enumerate(json_teams):
            tname = team.get('team', '?')
            players = team.get('players', [])
            if len(players) != 16:
                errors.append(f"{cat} team {i+1} '{tname}': expected 16 players, "
                              f"got {len(players)}")
            for label in ('team', 'country', 'coach'):
                bad = validate_string(team.get(label, ''), label)
                if bad:
                    errors.append(f"{cat} team {i+1} '{tname}' {label}: "
                                  f"invalid chars {bad!r} in '{team[label]}'")
            # Validate team-level attributes
            ctx = f"{cat} team {i+1} '{tname}'"
            err = _check_enum(team.get('tactic', '4-4-2'), TACTIC_VALUES, 7, f"{ctx}: tactic")
            if err:
                errors.append(err)
            skill = team.get('skill', 0)
            if not (0 <= skill <= 7):
                errors.append(f"{ctx}: skill must be 0-7, got {skill}")
            flag = team.get('flag', 0)
            if flag not in (0, 1):
                errors.append(f"{ctx}: flag must be 0 or 1, got {flag}")
            # Validate kit attributes
            kit = team.get('kit')
            if kit:
                for prefix in ('first', 'second'):
                    k = kit.get(prefix, {})
                    kctx = f"{ctx} {prefix} kit"
                    err = _check_enum(k.get('style', 'plain'), STYLE_VALUES, 3, f"{kctx}: style")
                    if err:
                        errors.append(err)
                    for field in ('shirt1', 'shirt2', 'shorts', 'socks'):
                        err = _check_enum(k.get(field, 'white'), COLOUR_VALUES, 15, f"{kctx}: {field}")
                        if err:
                            errors.append(err)
            # Validate player attributes
            for j, p in enumerate(players):
                if not isinstance(p, dict):
                    errors.append(f"{ctx} player {j+1}: "
                                  f"expected dict, got {type(p).__name__}")
                    continue
                bad = validate_string(p.get('name', ''), f"player {j+1}")
                if bad:
                    errors.append(f"{ctx} player {j+1}: "
                                  f"invalid chars {bad!r} in '{p['name']}'")
                num = p.get('number', 1)
                if not (1 <= num <= 16):
                    errors.append(f"{ctx} player {j+1}: "
                                  f"number must be 1-16, got {num}")
                pctx = f"{ctx} player {j+1}"
                for field, allowed, max_int, default in (
                    ('position', POSITION_VALUES, 15, 'goalkeeper'),
                    ('role', ROLE_VALUES, 3, 'goalkeeper'),
                    ('head', HEAD_VALUES, 2, 'white_dark'),
                ):
                    err = _check_enum(p.get(field, default), allowed, max_int, f"{pctx}: {field}")
                    if err:
                        errors.append(err)
            # Validate formation slot configuration
            starter_slots = []
            sub_count = 0
            for p in players:
                if not isinstance(p, dict):
                    continue
                pos = p.get('position', 'goalkeeper')
                if isinstance(pos, str):
                    pos = POSITION_VALUES.get(pos, -1)
                if pos == 15:
                    sub_count += 1
                else:
                    starter_slots.append(pos)
            if sorted(starter_slots) != list(range(11)):
                missing = set(range(11)) - set(starter_slots)
                duped = set(s for s in starter_slots if starter_slots.count(s) > 1)
                missing_names = [POSITION_NAMES.get(s, str(s)) for s in sorted(missing)]
                duped_names = [POSITION_NAMES.get(s, str(s)) for s in sorted(duped)]
                warnings.append(f"{ctx}: formation slots invalid — "
                                f"missing {missing_names}, duplicated {duped_names}")
            if sub_count != 5:
                warnings.append(f"{ctx}: expected 5 subs, got {sub_count}")

    return errors, warnings

File: frontend/src/lib/sslib/validate.ts
import {
  CHARSET, CATEGORIES,
  MAX_TEAM_NAME, MAX_COUNTRY, MAX_COACH, MAX_PLAYER_NAME,
  COLOUR_VALUES, STYLE_VALUES, HEAD_VALUES, ROLE_VALUES,
  POSITION_VALUES, POSITION_NAMES, TACTIC_VALUES,
} from './constants.js';
import { findPointerTable, chainWalkRegion } from './decode.js';
import type { TeamsJson } from './types.js';

export interface TeamErrors {
  team: string[];
  formation: string[];
  players: Record<number, string[]>;
}

export interface ValidationResult {
  valid: boolean;
  global: string[];
  teams: Record<string, Record<number, TeamErrors>>;
}

/** Precomputed ROM structure — expected team counts per category. */
export interface RomStructure {
  teamCounts: Record<string, number>;
}

/** Extract ROM structure once so validation doesn't rescan the ROM. */
export function extractRomStructure(romBytes: Uint8Array): RomStructure {
  const ptrs = findPointerTable(romBytes);
  const regionInfo = [
    ['national', ptrs.natStart, ptrs.natEnd],
    ['club', ptrs.clubStart, ptrs.clubEnd],
    ['custom', ptrs.custStart, ptrs.custEnd],
  ] as const;

  const teamCounts: Record<string, number> = {};
  for (const [cat, start, end] of regionInfo) {
    teamCounts[cat] = chainWalkRegion(romBytes, start, end).length;
  }
  return { teamCounts };
}

function validateString(text: string): string[] {
  const bad: string[] = [];
  for (const c of text.toUpperCase()) {
    if (!CHARSET.includes(c)) bad.push(c);
  }
  return bad;
}

function checkEnum(val: string | number, allowed: Record<string, number>, maxInt: number, label: string): string | null {
  if (typeof val === 'string') {
    if (!(val in allowed)) {
      return `${label} must be one of ${JSON.stringify(Object.keys(allowed).sort())}, got '${val}'`;
    }
  } else if (val < 0 || val > maxInt) {
    return `${label} must be 0-${maxInt}, got ${val}`;
  }
  return null;
}

export function validateTeams(romStructure: RomStructure, teamsJson: unknown): ValidationResult {
  const result: ValidationResult = { valid: true, global: [], teams: {} };

  if (
    typeof teamsJson !== 'object' || teamsJson === null ||
    !CATEGORIES.every(k => k in (teamsJson as Record<string, unknown>))
  ) {
    result.global.push("JSON must contain 'national', 'club', 'custom' keys");
    result.valid = false;
    return result;
  }

  const teams = teamsJson as TeamsJson;

  function getTeamErrors(cat: string, i: number): TeamErrors {
    if (!result.teams[cat]) result.teams[cat] = {};
    if (!result.teams[cat][i]) result.teams[cat][i] = { team: [], formation: [], players: {} };
    return result.teams[cat][i];
  }

  function addTeam(cat: string, i: number, msg: string) {
    getTeamErrors(cat, i).team.push(msg);
    result.valid = false;
  }

  function addFormation(cat: string, i: number, msg: string) {
    getTeamErrors(cat, i).formation.push(msg);
    result.valid = false;
  }

  function addPlayer(cat: string, i: number, j: number, msg: string) {
    const te = getTeamErrors(cat, i);
    if (!te.players[j]) te.players[j] = [];
    te.players[j].push(msg);
    result.valid = false;
  }

  for (const cat of CATEGORIES) {
    const romCount = romStructure.teamCounts[cat] ?? 0;
    const jsonTeams = teams[cat];
    if (jsonTeams.length !== romCount) {
      result.global.push(`${cat}: expected ${romCount} teams, got ${jsonTeams.length}`);
      result.valid = false;
      continue;
    }
    for (let i = 0; i < jsonTeams.length; i++) {
      const team = jsonTeams[i];
      const players = team.players ?? [];
      if (players.length !== 16) {
        addTeam(cat, i, `expected 16 players, got ${players.length}`);
      }
      const textMaxLens: Record<string, number> = { team: MAX_TEAM_NAME, country: MAX_COUNTRY, coach: MAX_COACH };
      for (const label of ['team', 'country', 'coach'] as const) {
        const val = team[label] ?? '';
        const bad = validateString(val);
        if (bad.length) {
          addTeam(cat, i, `${label}: invalid chars ${JSON.stringify(bad)} in '${val}'`);
        }
        if (val.length > textMaxLens[label]) {
          addTeam(cat, i, `${label}: max ${textMaxLens[label]} characters, got ${val.length}`);
        }
      }

      let err = checkEnum(team.tactic ?? '4-4-2', TACTIC_VALUES, 7, 'tactic');
      if (err) addTeam(cat, i, err);
      if (team.skill < 0 || team.skill > 7) addTeam(cat, i, `skill must be 0-7, got ${team.skill}`);
      if (team.flag !== 0 && team.flag !== 1) addTeam(cat, i, `flag must be 0 or 1, got ${team.flag}`);

      if (team.kit) {
        for (const prefix of ['first', 'second'] as const) {
          const k = team.kit[prefix] ?? {};
          err = checkEnum(k.style ?? 'plain', STYLE_VALUES, 3, `${prefix} kit style`);
          if (err) addTeam(cat, i, err);
          for (const field of ['shirt1', 'shirt2', 'shorts', 'socks'] as const) {
            err = checkEnum(k[field] ?? 'white', COLOUR_VALUES, 15, `${prefix} kit ${field}`);
            if (err) addTeam(cat, i, err);
          }
        }
      }

      for (let j = 0; j < players.length; j++) {
        const p = players[j];
        if (typeof p !== 'object' || p === null) {
          addPlayer(cat, i, j, `expected object, got ${typeof p}`);
          continue;
        }
        const name = p.name ?? '';
        const bad = validateString(name);
        if (bad.length) {
          addPlayer(cat, i, j, `invalid chars ${JSON.stringify(bad)} in '${name}'`);
        }
        if (name.length > MAX_PLAYER_NAME) {
          addPlayer(cat, i, j, `name: max ${MAX_PLAYER_NAME} characters, got ${name.length}`);
        }
        if (p.number < 1 || p.number > 16) {
          addPlayer(cat, i, j, `number must be 1-16, got ${p.number}`);
        }
        for (const [field, allowed, maxInt, def] of [
          ['position', POSITION_VALUES, 15, 'goalkeeper'],
          ['role', ROLE_VALUES, 3, 'goalkeeper'],
          ['head', HEAD_VALUES, 2, 'white_dark'],
        ] as [string, Record<string, number>, number, string][]) {
          err = checkEnum((p as Record<string, string | number>)[field] ?? def, allowed, maxInt, field);
          if (err) addPlayer(cat, i, j, err);
        }
      }

      // Validate formation slot configuration
      const starterSlots: number[] = [];
      const playerSlots: { pos: number; idx: number }[] = [];
      let subCount = 0;
      for (let j = 0; j < players.length; j++) {
        const p = players[j];
        if (typeof p !== 'object' || p === null) continue;
        const posRaw = p.position ?? 'goalkeeper';
        const pos = typeof posRaw === 'string' ? (POSITION_VALUES[posRaw] ?? -1) : posRaw as number;
        if (pos === 15) {
          subCount++;
        } else {
          starterSlots.push(pos);
          playerSlots.push({ pos, idx: j });
        }
      }
      const expected = Array.from({ length: 11 }, (_, k) => k);
      const sortedSlots = [...starterSlots].sort((a, b) => a - b);
      if (JSON.stringify(sortedSlots) !== JSON.stringify(expected)) {
        const missingSet = new Set(expected);
        starterSlots.forEach(s => missingSet.delete(s));
        const dupedSet = new Set(starterSlots.filter((s, idx, a) => a.indexOf(s) !== idx));
        const missingNames = [...missingSet].sort((a, b) => a - b).map(s => POSITION_NAMES[s] ?? String(s));
        const dupedNames = [...dupedSet].sort((a, b) => a - b).map(s => POSITION_NAMES[s] ?? String(s));
        addFormation(cat, i, `formation slots invalid — missing ${JSON.stringify(missingNames)}, duplicated ${JSON.stringify(dupedNames)}`);

        // Flag individual players that occupy duplicated positions
        for (const { pos, idx: j } of playerSlots) {
          if (dupedSet.has(pos)) {
            addPlayer(cat, i, j, `duplicate position: ${POSITION_NAMES[pos] ?? String(pos)}`);
          }
        }
      }
      if (subCount !== 5) {
        addFormation(cat, i, `expected 5 subs, got ${subCount}`);
      }
    }
  }

  // Clean up empty entries
  for (const cat of Object.keys(result.teams)) {
    for (const idx of Object.keys(result.teams[cat])) {
      const te = result.teams[cat][Number(idx)];
      if (te.team.length === 0 && te.formation.length === 0 && Object.keys(te.players).length === 0) {
        delete result.teams[cat][Number(idx)];
      }
    }
    if (Object.keys(result.teams[cat]).length === 0) {
      delete result.teams[cat];
    }
  }

  return result;
}
